File: Flutter/HUMAN_Demo/android/app/src/main/kotlin/com/humansecurity/demo/HumanManager.kt
package com.perimeterx.perimeterx

import android.app.Application
import android.os.Handler
import android.os.Looper
import com.humansecurity.mobile_sdk.HumanSecurity
import com.humansecurity.mobile_sdk.main.HSBotDefenderChallengeResult
import com.humansecurity.mobile_sdk.main.HSBotDefenderDelegate
import com.humansecurity.mobile_sdk.main.HSHybridChannel
import com.humansecurity.mobile_sdk.main.HSHybridSyncFailureReason
import com.humansecurity.mobile_sdk.main.HSHybridSyncStateKind
import com.humansecurity.mobile_sdk.main.policy.HSAutomaticInterceptorType
import com.humansecurity.mobile_sdk.main.policy.HSChallengePresentationType
import com.humansecurity.mobile_sdk.main.policy.HSPolicy
import com.humansecurity.mobile_sdk.main.policy.HSStorageMethod
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

class HumanManager {

    companion object {
        private var hasStarted = false
        private val mainHandler = Handler(Looper.getMainLooper())
        var eventSink: EventChannel.EventSink? = null

        private val delegate = object : HSBotDefenderDelegate {
            override fun botDefenderRequestBlocked(url: String?, appId: String) {}
            override fun botDefenderChallengeSolved(appId: String) {}
            override fun botDefenderChallengeCancelled(appId: String) {}
            override fun botDefenderChallengeRendered(appId: String) {}
            override fun botDefenderChallengeRenderFailed(appId: String) {}
            override fun botDefenderDidUpdateHeaders(headers: HashMap<String, String>, appId: String) {}

            override fun onHybridSyncFailed(
                appId: String,
                host: String,
                reason: HSHybridSyncFailureReason,
                isRecoverable: Boolean
            ) {
                emit(
                    mapOf(
                        "event" to "hybridSyncDidFail",
                        "appId" to appId,
                        "host" to host,
                        "reason" to reason.wireValue,
                        "isRecoverable" to isRecoverable
                    )
                )
            }

            override fun onHybridSyncRecovered(appId: String, host: String, channel: HSHybridChannel) {
                emit(
                    mapOf(
                        "event" to "hybridSyncDidRecover",
                        "appId" to appId,
                        "host" to host,
                        "channel" to channel.wireValue
                    )
                )
            }
        }

        fun handleEvent(application: Application, call: MethodCall, result: MethodChannel.Result) {
            try {
                when (call.method) {
                    "start" -> {
                        val appId = call.argument<String>("appId")
                        if (appId.isNullOrEmpty()) {
                            result.error("NO_APP_ID", "appId is required", null)
                            return
                        }
                        val domains = call.argument<List<String>>("webRootDomains")?.toSet() ?: emptySet()
                        val external = call.argument<Boolean>("supportExternalWebViews") ?: true
                        start(application, appId, domains, external)
                        result.success(null)
                    }
                    "vid" -> {
                        val appId = call.argument<String>("appId")
                        result.success(HumanSecurity.vid(appId))
                    }
                    "hybridSyncState" -> {
                        val appId = call.argument<String>("appId")
                        result.success(stateMap(HumanSecurity.hybridSyncState(appId?.takeIf { it.isNotEmpty() })))
                    }
                    "setupWebView" -> {
                        if (!hasStarted) {
                            result.error("NOT_STARTED", "Call HumanSecurity.start before setupWebView", null)
                            return
                        }
                        result.success(null)
                    }
                    "_getHumanHeaders" -> {
                        val json = JSONObject(HumanSecurity.BD.headersForURLRequest(null) as Map<*, *>?)
                        result.success(json.toString())
                    }
                    "_handleHumanResponse" -> {
                        val handled = HumanSecurity.BD.handleBlockResponse(call.arguments!! as String, null) { challengeResult: HSBotDefenderChallengeResult ->
                            result.success(if (challengeResult == HSBotDefenderChallengeResult.SOLVED) "solved" else "cancelled")
                            null
                        }
                        if (!handled) {
                            result.success("false")
                        }
                    }
                    else -> result.notImplemented()
                }
            } catch (exception: Exception) {
                result.error("HUMAN_ERROR", exception.message, null)
            }
        }

        private fun start(
            application: Application,
            appId: String,
            domains: Set<String>,
            supportExternalWebViews: Boolean
        ) {
            if (hasStarted) {
                return
            }
            val policy = HSPolicy()
            policy.storageMethod = HSStorageMethod.DATA_STORE
            policy.automaticInterceptorPolicy.interceptorType = HSAutomaticInterceptorType.NONE
            policy.doctorAppPolicy.enabled = true
            policy.challengePresentationType = HSChallengePresentationType.CENTER
            policy.hybridAppPolicy.supportExternalWebViews = supportExternalWebViews
            if (domains.isNotEmpty()) {
                policy.hybridAppPolicy.setWebRootDomains(domains, appId)
            }
            HumanSecurity.start(application, appId, policy)
            HumanSecurity.BD.delegate = delegate
            hasStarted = true
        }

        private fun stateMap(state: com.humansecurity.mobile_sdk.main.HSHybridSyncState): Map<String, Any> {
            val kind = when (state.kind) {
                HSHybridSyncStateKind.NOT_APPLICABLE -> "notApplicable"
                HSHybridSyncStateKind.PENDING -> "pending"
                HSHybridSyncStateKind.HEALTHY -> "healthy"
                HSHybridSyncStateKind.DEGRADED -> "degraded"
                HSHybridSyncStateKind.FAILED -> "failed"
            }
            val includeChannel = state.kind == HSHybridSyncStateKind.HEALTHY ||
                state.kind == HSHybridSyncStateKind.DEGRADED
            return mapOf(
                "kind" to kind,
                "isHealthy" to state.isHealthy,
                "channel" to if (includeChannel) state.channel?.wireValue ?: "" else "",
                "reason" to if (state.kind == HSHybridSyncStateKind.FAILED) state.reason?.wireValue ?: "" else ""
            )
        }

        private fun emit(payload: Map<String, Any>) {
            mainHandler.post {
                try {
                    eventSink?.success(payload)
                } catch (_: Exception) {
                }
            }
        }
    }
}

File: Flutter/HUMAN_Demo/ios/HumanDemo/HumanManager.swift
import Foundation
import Flutter
import HUMAN

class HumanManager: NSObject, HSBotDefenderDelegate, FlutterStreamHandler {
    static let shared = HumanManager()

    private var hasStarted = false
    private var eventSink: FlutterEventSink?

    func start(appId: String, domains: [String], supportExternalWebViews: Bool) throws {
        if hasStarted {
            return
        }
        let policy = HSPolicy()
        policy.automaticInterceptorPolicy.interceptorType = .none
        policy.doctorAppPolicy.enabled = true
        policy.hybridAppPolicy.supportExternalWebViews = supportExternalWebViews
        if !domains.isEmpty {
            policy.hybridAppPolicy.set(webRootDomains: Set(domains), forAppId: appId)
        }
        try HumanSecurity.start(appId: appId, policy: policy)
        HumanSecurity.BD.delegate = self
        hasStarted = true
    }

    func setupChannel(with controller: FlutterViewController) {
        let messenger = controller.binaryMessenger
        let humanChannel = FlutterMethodChannel(name: "com.humansecurity.demo/human", binaryMessenger: messenger)
        humanChannel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call, result: result)
        }
        let events = FlutterEventChannel(name: "com.humansecurity.demo/human_events", binaryMessenger: messenger)
        events.setStreamHandler(self)
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    func hybridSyncDidFail(appId: String, host: String, reason: HSHybridSyncFailureReason, isRecoverable: Bool) {
        eventSink?([
            "event": "hybridSyncDidFail",
            "appId": appId,
            "host": host,
            "reason": reason.wireValue,
            "isRecoverable": isRecoverable
        ])
    }

    func hybridSyncDidRecover(appId: String, host: String, channel: HSHybridChannel) {
        eventSink?([
            "event": "hybridSyncDidRecover",
            "appId": appId,
            "host": host,
            "channel": channel.wireValue
        ])
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start":
            let args = call.arguments as? [String: Any]
            let appId = args?["appId"] as? String ?? ""
            let domains = args?["webRootDomains"] as? [String] ?? []
            let external = args?["supportExternalWebViews"] as? Bool ?? true
            if appId.isEmpty {
                result(FlutterError(code: "NO_APP_ID", message: "appId is required", details: nil))
                return
            }
            DispatchQueue.main.async {
                do {
                    try self.start(appId: appId, domains: domains, supportExternalWebViews: external)
                    result(nil)
                } catch {
                    result(FlutterError(code: "START_ERROR", message: "\(error)", details: nil))
                }
            }
        case "vid":
            let args = call.arguments as? [String: Any]
            let appId = args?["appId"] as? String
            result(HumanSecurity.vid(forAppId: appId))
        case "hybridSyncState":
            let args = call.arguments as? [String: Any]
            let appId = args?["appId"] as? String
            let resolved = (appId?.isEmpty == false) ? appId : nil
            result(Self.dictionary(for: HumanSecurity.hybridSyncState(forAppId: resolved)))
        case "setupWebView":
            if !hasStarted {
                result(FlutterError(code: "NOT_STARTED", message: "Call HumanSecurity.start before setupWebView", details: nil))
                return
            }
            result(nil)
        case "_getHumanHeaders":
            var json: String?
            do {
                let headers = HumanSecurity.BD.headersForURLRequest(forAppId: nil)
                let data = try JSONSerialization.data(withJSONObject: headers)
                json = String(data: data, encoding: .utf8)
            } catch {
                print("error: \(error)")
            }
            result(json)
        case "_handleHumanResponse":
            if let response = call.arguments as? String, let data = response.data(using: .utf8) {
                let handled = HumanSecurity.BD.handleBlockResponse(data: data, url: nil) { challengeResult in
                    result(challengeResult == .solved ? "solved" : "cancelled")
                }
                if handled {
                    return
                }
            }
            result("false")
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private static func dictionary(for state: HSHybridSyncState) -> [String: Any] {
        let kind: String
        switch state.kind {
        case .notApplicable: kind = "notApplicable"
        case .pending: kind = "pending"
        case .healthy: kind = "healthy"
        case .degraded: kind = "degraded"
        case .failed: kind = "failed"
        }
        var channel = ""
        var reason = ""
        if state.hasChannel, let value = state.channel {
            channel = value.wireValue
        }
        if state.hasReason, let value = state.reason {
            reason = value.wireValue
        }
        return [
            "kind": kind,
            "isHealthy": state.isHealthy,
            "channel": channel,
            "reason": reason
        ]
    }
}
